generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -49,71 +52,121 @@ case object GlobalResetAsynchronousFull extends GlobalResetScheme with HasAsyncI
 case object GlobalResetSchemeKey extends Field[GlobalResetScheme](GlobalResetSynchronous)
 
 
+
 /**
  * The base class used for building chips. This constructor instantiates a module specified by the BuildSystem parameter,
  * named "system", which is an instance of DigitalTop by default. The default clock and reset for "system" are set by two
  * wires, "systemClock" and "systemReset", which are intended to be driven by traits mixed-in with this base class.
  */
-abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with HasTestHarnessFunctions {
+abstract class BaseChipTop(implicit p: Parameters) extends LazyModule with HasTestHarnessFunctions {
 
   // A publicly accessible list of IO cells (useful for a floorplanning tool, for example)
   val iocells = ArrayBuffer.empty[IOCell]
   // A list of functions to call in the test harness
   val harnessFunctions = ArrayBuffer.empty[TestHarnessFunction]
-  // The system clock
-  // These are given so that IOCell can use DataMirror and generate ports with
-  // the right flow (Input/Output)
-  val systemClock = Wire(Input(Clock()))
-  val systemReset = Wire(Input(Reset()))
 
   // The system module specified by BuildSystem
   val lSystem = p(BuildSystem)(p).suggestName("system")
-  val system = withClockAndReset(systemClock, systemReset) { Module(lSystem.module) }
-
-  // Call all of the IOBinders and provide them with a default clock and reset
-  withClockAndReset(systemClock, systemReset) {
-    // Call each IOBinder on both the lazyModule instance and the module
-    // instance. Generally, an IOBinder PF should only be defined on one, so
-    // this should not lead to two invocations.
-    val (_ports, _iocells, _harnessFunctions) = p(IOBinders).values.flatMap(f => f(lSystem) ++ f(system)).unzip3
-    // We ignore _ports for now...
-    iocells ++= _iocells.flatten
-    harnessFunctions ++= _harnessFunctions.flatten
-  }
 
+  // The provider of all clock groups. Subclasses define this
+  def provideDiplomaticClocks(): Unit
+
+  // This function generates RTL that may be necessary to drive the diplomatic clocks
+  // Subclasses define this
+  def driveClockGroups(systemClock: Clock, systemReset: Reset): (Seq[IOCell], Seq[TestHarnessFunction])
```

Comment:
I wonder if we often will need to use systemClock and systemReset in this function.
It seems like a recipe for many self-inflicted wounds where someone uses systemClock or more insidiously systemReset to drive the diplomatic clocks including what will become systemClock/Reset.

Reply:
Yeah this is pretty confusing already. It looks like in this example the `ResetCatchAndSync` can be clocked to the incoming clock from the TestHarness. 
It also seems like reset-generation can be refactored to be more elegantly tied to clock-generation.